### DISCUSSION TITLE
Livewire loosing $this with nested components

### DISCUSSION THREAD

**Discussion by tanthammar** [livewire/livewire#2313] [1. Help] [answered]
I'm trying to figure out why Livewire looses `$this` when nesting child components.

I'm the author of tall-forms pkg, a Livewire form-generator.
All the fields in that package are Blade views/classes except one that I had to make as a separate Livewire component.

What I don't understand is, why I can't use `$this` in blade views, for fields that comes **after** the nested Livewire component.
Error: `Using $this when not in object context`

Nor can I call any property like `$form_data`, in fields that comes after the child component.
Error: `Undefined variable`

Conceputal example.
 
```blade
<div>//parent component

// these will work fine
    <x-field-foo :field="$field" :value="data_get($this, $field->key)" />
    <x-field-bar :field="$field" :value="data_get($this, $field->key)" />

// here is a nested component (works)
    <livewire:foo-component :model="$model" :field="$field->fieldToArray()" :tags="data_get($form_data, $field->name, '')" />

// all fields after the nested component gives error; Using $this when not in object context
    <x-field-baz :field="$field" :value="data_get($this, $field->key)" />

// or calling a property gives error; Undefined variable [property name], in this case $form_data
   <x-field-foo :field="$field" :value="data_get($form_data, 'bar.baz')" />

</div>
```
 
Putting the nested Livewire component last, resolves the issue.

Is there something conceptual I'm missing?

**Comment by joshhanley** [comment]
@tanthammar I tried to replicate it but I couldn't get the issue to happen, it all seemed to work ok. Maybe I'm missing something. Any chance you can provide a test repo?

This is what I tried (running on v2.3.6)
```blade
<div> // Parent component
        <div>
            {{ $testing }}
        </div>

        <livewire:foo-component />

        <div>{{ data_get($this, 'testing') }}</div>
        <div>{{ data_get($sample, 'test') }}</div>
</div>
```

```php
    // Parent component
    public $testing = "test";

    public $sample = [
        'test' => 'some new test'
    ];
```

```blade
<div>
    Foo component {{ data_get($this, 'foo') }}
</div>
```

```php
// Foo component
public $foo = "bob";
```

**Reply by christophheich** [reply]
@joshhanley 

This issue happens to me as well.

To reproduce this issue you will need to follow these steps:

1. Create a variable in the parent livewire component eg.: public bool $test = true;
2. Access the variable in the component blade file eg.: @if($this->test)
3. Add a nested livewire component somewhere to the parent blade file eg.: @livewire('randomComponent')
4. The issue should now be: Using $this when not in object context on page load

Would be appreciated if someone can confirm this.

**Reply by marispro** [reply]
I guess second step is wrong. To access variable in blade should use `@if($test)`

**Reply by christophheich** [reply]
> I guess second step is wrong. To access variable in blade should use `@if($test)`

The thing is, the variable is undefined if you are not using $this->test while trying to access it in a laravel blade component while $this->test works if there is no nested component in the parent blade.

**Reply by christophheich** [reply]
What i found out is, that this issue only happens when a livewire component has a blade component with a nested livewire component in its slot. 

```
random livewire component
<div>

    <h1>Random</h1>

     <x-random-blade-component>

                @livewire('random-nested-livewire-component')

        </x-random-blade-component>

</div>
```

calling $this->randomVariable in the random-blade-component works if there is no nested livewire component in the blade component slot, if the nested component is outside of the blade component slot it will continue to work as normal.

**Reply by imClement** [reply]
> What i found out is, that this issue only happens when a livewire component has a blade component with a nested livewire component in its slot.

I can confirm the issue in this context and also, if a named slot is used, too (instead of the default $slot)

```
<div>

    <h1>Random</h1>

     <x-random-blade-component>

              <x-slot name="header">

                @livewire('random-nested-livewire-component')

             </x-slot>


            Default Slot Content


        </x-random-blade-component>

</div>
```

**Comment by jeffchown** [comment]
@joshhanley I just started having the same problem.  I have a livewire/Dashboard component, in its dashboard.blade.php, @include multiple blade files some of which reference computed properties and/or methods in the Dashboard component via `$this->someComputedProperty`.  

Everything works fine *until* I add a nested livewire component.  Once I add the nested component to the `dashboard.blade.php`, any of the @include(d) blade file code that references the Dashboad component's computed properties/methods via `$this->...` now throw a `Using $this when not in object context` error.

One thing I was able to figure out that will hopefully help, is that the error is only thrown if the @include that contains the `$this->...` is *below* the call to the nested livewire component.

e.g. This will throw the error:
`livewire/dashboard.blade.php`:
```
<div>
  <livewire:some-nested-component />
  @include('partial-containing-this-usage.blade.php')
</div>
```

...and this will *not* throw the error:
```
<div>
  @include('partial-containing-this-usage.blade.php')
  <livewire:some-nested-component />
</div>
```

**Reply by jeffchown** [reply]
UPDATE: I found if I change the `@include` to pass the `$this` variable to the partial under a different name, e.g.:
`@include('partial-containing-this-usage.blade.php', ['parent' => $this])`

then, in the partial change my `$this->...` reference(s) to `$parent->...` everything works as expected.

Not ideal, is a workaround, but I also hope it helps lead to an explanation of what's happening and why and if it's something that can be fixed/changed.

**Reply by joshhanley** [reply]
@jeffchown thanks for the extra info! Any chance you can provide a sample repo so I can pull it and see if I can work out what's going on?

**Reply by christophheich** [reply]
For a blade component the following syntax will work: `<x-random-blade-component :infiniteScroll="$this->infiniteScroll">` while infiniteScroll is the variable that can be accessed in the blade component using `$infiniteScroll` and passing over the variable using `$this->infiniteScroll`

**Reply by christophheich** [reply]
> @jeffchown thanks for the extra info! Any chance you can provide a sample repo so I can pull it and see if I can work out what's going on?

There you go: https://github.com/christophheich/livewire-nested-blade

The resources folder contains the blade components (one using $this and one through passing the data from the livewire component to the blade component)

Regards

**Reply by joshhanley** [reply]
I have added a failing test for this in PR #2786.

There is also a temporary work around listed in the PR, if anyone is having this issue.

Hope this helps!

**Comment by joshhanley** [comment, marked as answer]
I have added a failing test for this in PR #2786.

There is also a temporary work around listed in the PR, if anyone is having this issue.

Hope this helps!

**Reply by tanthammar** [reply]
🎉 Confirmed!
This solves #1653, #1401, #2380, #2313

@joshhanley
🌟 You are a lifesaver!!!!!!
Amazed by your perseverance to fix this and other issues with Livewire as well as your politeness to the community.

@calebporzio
This fix is actually a game changer for Livewire.
I hope you give @joshhanley a real treat for solving this, (and other) issues :)
It held me and others back to create more advanced Livewire components that requires nesting mixed with blade components.

Now I'll go out on my balcony and do some crazy jumping while shouting out to the world, that the lid finally popped off the Livewire jar 💖 (Then I'll remove some nasty hacks from the tall-forms package that worked around this issue.)

Sending LOVE!

**Reply by jeffchown** [reply]
Thanks @joshhanley !  

I *loudly* echo @tanthammar 's kudos and thanks!  This is indeed a game-changer and resolves multiple issues on my end as well.
